armory/art_experimental/attacks/kenansville_dft.py

```diff
@@ -42,56 +41,26 @@ def __init__(
             raise ValueError("Negative SNR is not allowed")
 
     def _attack(self, x):
-        x_len = len(x)
+        # Scale the threshold based on the target SNR
+        threshold = 10 ** (-self.snr_db / 10)
         x_fft = np.fft.fft(x)
         x_psd = np.abs(x_fft) ** 2
-        # sort by frequencies with increasing power
+        # Scale the threshold based on the strenght of the signal
+        threshold *= np.sum(x_psd)
+        # Sort frequencies by amplitude
         x_psd_ind = np.argsort(x_psd)
         dc_ind = np.where(x_psd_ind == 0)[0][0]
-        signal_db = 10 * np.log10(np.sum(x_psd))
-
-        """
-        The goal of the following search is to find all the
-        low power frequencies that can be discarded while
-        maintaining a minimum SNR.
-
-        If desired, the following coarse and fine search could
-        be replaced with binary search for faster convergence.
-        """
-        # coarse search
-        id = 2
-        noise = np.sum(x_psd[x_psd_ind[:id]])
-        noise_db = 10 * np.log10(noise)
-        while signal_db - noise_db > self.snr_db:
-            id *= 2
-            noise = np.sum(x_psd[x_psd_ind[: min(id, x_len)]])
-            noise_db = 10 * np.log10(noise)
-
-        if id == 2:
-            return x
-
-        # fine search
-        id = int(id / 2)
-        noise = np.sum(x_psd[x_psd_ind[:id]])
-        noise_db = 10 * np.log10(noise)
-        while signal_db - noise_db > self.snr_db:
-            id += 2
-            noise = np.sum(x_psd[x_psd_ind[: min(id, x_len)]])
-            noise_db = 10 * np.log10(noise)
-
-        id -= 2
-
+        reordered = x_psd[x_psd_ind]
+        # Compute the cumulative perturbation size induced by zeroing frequencies up to index ix
+        # Then searching the first one that is below the threshold
+        id = np.searchsorted(np.cumsum(reordered), threshold)
+        id -= id % 2
```

Comment:
In the initial implementation, `id` is guaranteed to be even immediately after the fine search, but with the change to the search, I believe the `id` returned by the search can now be odd.  Does this require any changes to how `id` is modified after this point? In particular, if `len(x)` is odd and `dc_ind < id`, do we still need to set `id -= id % 2`?